### PR TITLE
DAOS-12554 bio: BIO unit tests for WAL

### PR DESCRIPTION
**Pull request by NiuYawei**
Added BIO unit tests to test WAL APIs, fixed two defects:
- defect in get_trans_blk(), triggered when WAL log wraps;
- defect in fill_trans_blks(), triggered for UMEM_ACT_MOVE action;

This pach also removed bio_wal_ckp_start() and renamed bio_wal_ckp_end()
to bio_wal_checkpoint().

Required-githooks: true

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/bio/tests/wal_ut.c [hunk 1]
```diff
@@ -0,0 +1,815 @@
...
+
+	return fake_wal_tx;
+}
+
+static inline uint64_t
+rand_addr(void)
+{
+	uint64_t addr  = rand();
+
+	return (addr << 32) + rand();
```
Comment (wangshilong) [inline src/bio/tests/wal_ut.c:149]: maybe just addr << 32 + addr is fine.

**Review by wangshilong — DISMISSED** [review]

File: src/bio/tests/bio_ut.c [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	if (rc != 0)
+		return rc;
+
+	rc = ABT_init(0, NULL);
+	if (rc != 0)
+		goto out_debug;
+
+	fd = open(nvme_conf, O_RDONLY, 0600);
+	if (fd < 0) {
+		D_ERROR("Failed to open %s. %s\n", nvme_conf, strerror(errno));
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/bio/tests/bio_ut.c:53]: indeed, thanks.

File: src/bio/tests/bio_ut.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+#define D_LOGFAC	DD_FAC(tests)
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <getopt.h>
+#include <fcntl.h>
+#include "bio_ut.h"
+
+static char		db_path[50];
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/bio/tests/bio_ut.c:15]: ok, let's bump it to 100.

File: src/bio/tests/wal_ut.c [hunk 4]
```diff
@@ -0,0 +1,815 @@
...
+		ut_tx_add_action(tx, UMEM_ACT_SET);
+		ut_tx_add_action(tx, UMEM_ACT_SET_BITS);
+		ut_tx_add_action(tx, UMEM_ACT_CLR_BITS);
+
+		rc = bio_wal_reserve(args->bua_mc, &tx->utx_id);
+		assert_rc_equal(rc, 0);
+
+		rc = bio_wal_commit(args->bua_mc, tx, NULL);
+		assert_rc_equal(rc, 0);
+	}
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/bio/tests/wal_ut.c:619]: yeah, I used same "actions" in these two tests, actually I should use different "actions".

File: src/bio/tests/wal_ut.c (hunk 1 above)
Comment (NiuYawei) [reply src/bio/tests/wal_ut.c:149]: well, use two different rand number looks better to me.

**Review by wangshilong — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**